[aekyr] iP

lucasisnotcool opened the pull request:
# L.A.F.F.Y
> "Hello, I am Laffy, Lucas' Automated Friendship For You." - Laffy 😊

Laffy is a task management application that helps you keep track of your tasks. It is a CLI app that allows you to add, mark, unmark, list, delete, and find tasks.

## Using Laffy
1. download the latest release from [here](https://github.com/aekyr/ip/releases)
2. Run the jar file using the command `java -jar laffy.jar`
3. Start chatting with Laffy!

## Features

- **Add tasks:** Add different types of tasks such as todos, deadlines, and events.
- **Mark tasks:** Mark tasks as done.
- **Unmark tasks:** Unmark tasks that are marked as done.
- **List tasks:** List all tasks.
- **Delete tasks:** Delete tasks from the list.
- **Find tasks:** Find tasks by searching for keywords.

## Commands
1. `list`: Lists all tasks.
2. `todo <description>`: Adds a todo task to the list.
3. `deadline <description> /by <date>`: Adds a deadline task to the list.
4. `event <description> /from <date> /to <date>`: Adds an event task to the list.
5. `mark <index>`: Marks a task as done.
6. `unmark <index>`: Marks a task as undone.
7. `delete <index>`: Deletes a task from the list.
8. `find <keyword>`: Finds tasks that contain the keyword.
9. `bye`: Exits the application.
10. `help`: Displays the list of commands.

### To do list

- [x] ~~Create project structure~~
- [x] ~~Implement basic features~~
- [ ] Add **_friendship_**
- [ ] Write documentation

### Main Function of Laffy Application (For those who would like to make Laffy your own)

```java
public static void main(String[] args) throws IOException {
    new Laffy("data/laffy.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

lucasisnotcool committed:
Level 0. Rename, Greet, Exit
(a) Give your chatbot a new name, to differentiate it from the placeholder name Duke.
Avoid these common choices as well: Chatty, Jarvis, ChatBot, Chad

(b) Implement an initial skeletal version of it that simply greets the user and exits.

lucasisnotcool committed:
Improve the skeletal version of Duke so that it echos commands entered by the user, and exits when the user types the command bye.

lucasisnotcool committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested.

lucasisnotcool committed:
Add the ability to mark tasks as done. Optionally, add the ability to change the status back to not done.

lucasisnotcool committed:
ToDos, Events, Deadlines
Add support for tracking three types of tasks:

ToDos: tasks without any date/time attached to it e.g., visit new theme park
Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
Events: tasks that start at a specific date/time and ends at a specific date/time
e.g., (a) team project meeting 2/10/2019 2-4pm (b) orientation week 4/10/2019 to 11/10/2019

lucasisnotcool committed:
Shifted command handling to its own class and subclasses

lucasisnotcool committed:
Test using the I/O redirection technique

Use the input/output redirection technique to semi-automate the testing of Duke.

lucasisnotcool committed:
Handle Errors
Teach the chatbot to deal with errors such as incorrect inputs entered by the user.

lucasisnotcool committed:
Added DeleteCommand. Delete
Add support for deleting tasks from the list.

lucasisnotcool committed:
Refactored code for neatness

lucasisnotcool committed:
removed deprecated methods in Laffy

lucasisnotcool committed:
Change InvalidCommand getusage to use class field instead

lucasisnotcool committed:
refactor class tasklist to match java conventions

lucasisnotcool committed:
de-packaged tasklist since command parser needs to be its own package in future

lucasisnotcool committed:
add file read and create

lucasisnotcool committed:
Add save feature to file at data/laffy.txt

lucasisnotcool committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add save feature to file at data/laffy.txt
  add file read and create
  de-packaged tasklist since command parser needs to be its own package in future
  refactor class tasklist to match java conventions
  Change InvalidCommand getusage to use class field instead

lucasisnotcool committed:
Implement datetimes to deadline and event classes for by, from and to fields

lucasisnotcool committed:
Update tests for new deadline and event commands

lucasisnotcool committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Update tests for new deadline and event commands
  Implement datetimes to deadline and event classes for by, from and to fields

lucasisnotcool committed:
Refactor ListCommandParser to just Parser

lucasisnotcool committed:
Add storage class to encapsulate storage operations

lucasisnotcool committed:
Extracted Ui class

lucasisnotcool committed:
Fixed bug where file was not saving

lucasisnotcool committed:
Change command class to act upon tasklist storage ui instead of having tasklist storage ui

lucasisnotcool committed:
Put everything into single laffy package

lucasisnotcool committed:
Package command into own package

lucasisnotcool committed:
Implement exception throwing for commands and exception handling for ui

lucasisnotcool committed:
Package tasks together in the tasklist package

lucasisnotcool committed:
Merge remote-tracking branch 'origin/add-gradle-support'

* origin/add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

lucasisnotcool committed:
Add gradle to project and use it to build

lucasisnotcool committed:
Add tests for tasklist.task and tasklist.taskdateapi

lucasisnotcool committed:
Package app as jar file with gradlew shadowJar

lucasisnotcool committed:
Shift ui call to command's responsibility instead of run

lucasisnotcool committed:
Clean up java code in line with coding standard

lucasisnotcool committed:
Add javadocs for most methods and constructors that are non-trivial

lucasisnotcool committed:
Add find command

lucasisnotcool committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add javadocs for most methods and constructors that are non-trivial

lucasisnotcool committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Clean up java code in line with coding standard
  Shift ui call to command's responsibility instead of run

# Conflicts:
#	src/main/java/laffy/Storage.java
#	src/main/java/laffy/command/AddDeadlineCommand.java
#	src/main/java/laffy/command/AddEventCommand.java
#	src/main/java/laffy/command/AddTodoCommand.java
#	src/main/java/laffy/command/Command.java
#	src/main/java/laffy/command/DeleteCommand.java
#	src/main/java/laffy/command/ExitCommand.java
#	src/main/java/laffy/command/IndexedCommand.java
#	src/main/java/laffy/command/ListCommand.java
#	src/main/java/laffy/command/MarkCommand.java
#	src/main/java/laffy/command/UnmarkCommand.java

lucasisnotcool committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/laffy/Parser.java
#	src/main/java/laffy/tasklist/TaskList.java

lucasisnotcool committed:
Add header comments for classes

lucasisnotcool committed:
Swap commandword implementation to enums

lucasisnotcool committed:
Add checkstyle and fix coding style violations

lucasisnotcool committed:
Merge branch 'branch-A-Checkstyle'

* branch-A-Checkstyle:
  Add checkstyle and fix coding style violations

lucasisnotcool committed:
add basic gui that replies what the user sent

lucasisnotcool committed:
Seperate view and controller using javafx and fxml

lucasisnotcool committed:
Style gui to look nicer

lucasisnotcool committed:
Refactor Commands to send String output for gui while keeping CLI ui functionality.

lucasisnotcool committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Refactor Commands to send String output for gui while keeping CLI ui functionality.
  Style gui to look nicer
  Seperate view and controller using javafx and fxml
  add basic gui that replies what the user sent

hutongyan commented on src/main/java/laffy/Laffy.java:
```diff
@@ -0,0 +1,67 @@
+            System.out.print("> ");
+            try {
+                String fullCommand = ui.readCommand();
+                ui.showLine(); // show the divider line ("_______")
```
Suggest to remove the in-line //

lucasisnotcool replied:
got it 😄

hutongyan commented on src/main/java/laffy/Storage.java:
```diff
@@ -0,0 +1,142 @@
+        loadData();
+    }
+
+    private static boolean isValidType(String type) {
+        return type.equals("T") || type.equals("D") || type.equals("E");
+    }
+
+    private static boolean isValidDone(String done) {
+        return done.equals("0") || done.equals("1");
+    }
+
+    private static boolean isValidDesc(String desc) {
+        return !desc.isBlank() && !desc.isEmpty();
+    }
+
+    private static boolean isValidBy(String by) {
+        return !by.isBlank() && !by.isEmpty();
+    }
+
+    private static boolean isValidFrom(String from) {
+        return !from.isBlank() && !from.isEmpty();
+    }
+
+    private static boolean isValidTo(String to) {
+        return !to.isBlank() && !to.isEmpty();
+    }
```
should add javaDoc to these methods?

lucasisnotcool replied:
JavaDoc only required for public methods, thanks for the suggestion!

hutongyan commented on src/main/java/laffy/Ui.java:
```diff
@@ -0,0 +1,37 @@
+public class Ui {
+    private final Scanner sc;
+
+    public Ui() {
+        this.sc = new Scanner(System.in);
+    }
+
+    public String readCommand() {
+        return sc.nextLine();
+    }
+
+    public void showLine() {
+        System.out.println("____________________________________________________________");
+    }
+
+    public void greet() {
+        showLine();
+        System.out.println("Hello! I'm L.A.F.F.Y");
+        System.out.println("What can I do for you?");
+        showLine();
+    }
+
+    public void showError(String message) {
+        System.out.println(message);
+    }
+
+    public void echo(String input) {
+        System.out.println(input);
+    }
```
Should add javadoc to these methods?

lucasisnotcool replied:
Done! Thanks for the nice catch 🔥

hutongyan commented on src/main/java/laffy/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,77 @@
+        }
+    }
+
+    @Override
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws TaskListException {
+        ui.echo("Got it. I've added this task:\n  "
+                + taskList.addDeadline(desc, by)
+                + "\nNow you have " + taskList.size() + " tasks in the list.");
+        super.execute(taskList, ui, storage);
+    }
+
+
+    public static String getDescription() {
+        return getCommandWord() + " <description> /by <deadline: dd-MM-yy[ HHMM]>";
+    }
+
+    public String getUsage() {
+        return super.getUsage() + getDescription();
+    }
+
+    public static String getCommandWord() {
+        return COMMAND_WORD.toString();
+    }
```
Should add javadoc to these methods?

hutongyan commented on src/main/java/laffy/command/exceptions/MissingKeywordFlag.java:
```diff
@@ -0,0 +1,12 @@
+public class MissingKeywordFlag extends LaffyException {
+    private static final String message = "Missing keyword flag. ";
+
+    public MissingKeywordFlag(String explanation) {
```
Maybe method name not that appropriate?

lucasisnotcool replied:
Hmm this is the constructor for the Exception, so maybe it's ok 🤔

hutongyan reviewed (commented):
Overall, I found your code is easy to read, and follows the coding standard, except for some missing javadoc comment and some method names fail to follow the Camel standard.

lucasisnotcool committed:
Add assertions to validate task data format

The loadData method checks if each task line read from the file is valid
 but does not assert that the number of fields match the expected format
 for ToDo, Deadline, or Event tasks. This can lead to subtle bugs if the
 data is malformed.

Let's add assertions to confirm the data length for each task type.
Using assertions is a lightweight way to detect unexpected data formats
during development while avoiding a runtime performance impact in
production.

lucasisnotcool committed:
Add assertions for task fields and refactor to switch-case

The loadData method checks if each task line is valid before parsing,
but it does not assert that the type, done status, and description
fields are valid at the outset. Without assertions, invalid data might
slip through unnoticed, leading to subtle bugs.

Additionally, using multiple if-else statements to handle task types is
less readable. A switch-case structure is more straightforward and
reduces branching complexity.

Let's assert the validity of type, done, and desc right after reading
each task’s data.

Let's also replace the if-else statements with a switch-case to
simplify the code and make it easier to maintain.

lucasisnotcool committed:
Add assertion for matching fields count in saveData

The saveData method converts each task’s data into a pipe-separated
string but does not verify that the number of fields in the resulting
string matches the number of elements in the taskData list. This can
lead to silent data corruption if there is a mismatch.

Let's add an assertion checking that line.split(" \| ").length equals
taskData.size(). This ensures that the string representation of each
task is consistent with its internal data structure, helping detect any
discrepancies before writing to the file.

lucasisnotcool committed:
Merge pull request #1 from aekyr/branch-A-Assertions

Branch a assertions

lucasisnotcool committed:
Extract parseTaskDataLine method from LoadData, to adhere to method length requirement and SLAP

lucasisnotcool committed:
Alter switches to adhere to code quality guidelines for using switches

lucasisnotcool committed:
Merge pull request #2 from aekyr/branch-A-CodeQuality

Branch a code quality

lucasisnotcool committed:
Add upcoming command for deadlines and events

Fulfills the BCD change of B-Reminders.
Currently, there's no straightforward way to remind users about
approaching tasks. This makes it harder for users to prioritize based
on urgency.

A dedicated upcoming command provides a quick view of time-sensitive
tasks by sorting them from the most urgent to the least.

Let's implement the "soon" keyword, which lists all upcoming deadlines
(sorted by the 'by' field) and events (sorted by the 'from' field).
This helps users see at a glance which tasks require immediate attention
 and plan their schedules more effectively.

lucasisnotcool committed:
Merge branch 'branch-BCD-Extension'

* branch-BCD-Extension:
  Add upcoming command for deadlines and events

lucasisnotcool committed:
Add tests for Todo and TaskList

lucasisnotcool committed:
Add tests for Deadline

lucasisnotcool committed:
Add tests for Event

lucasisnotcool committed:
Fix checkstyle issues for tests

lucasisnotcool committed:
Add delete method test for tasklist

lucasisnotcool committed:
Add tests for toTaskData and sequential deletes in TaskList

lucasisnotcool committed:
Add user guide and product screenshot to docs

lucasisnotcool committed:
Update readme in root directory

lucasisnotcool committed:
Resolve issues related to code quality from peer review
